### DISCUSSION TITLE
More input event types

### DISCUSSION THREAD

**Discussion by fscherf** [lona-web-org/lona#179] [Ideas]
This idea or feature request is originally from @mwx23 (#169)

Currently Lona has the input event types ``CLICK`` and ``CHANGE``.
There are some ideas for new types

* ``FOCUS`` and ``BLUR``
  * These two are very simple to implement
* ``HOVER``
  * Hovering would have to be implemented with a timeout like the ``oninput`` events. The event would fire after the cursor remained on the node for a specific amount of time to avoid flooding the server with events, when the cursor gets moved over the page 
* Keyboard events
  * An API for keyboard events could look like this: ``Div(events=[SHIFT+ENTER])``
* Drag and Drop
  * I have no idea yet how to implement that and if it is a good idea. Maybe it makes more sense to create a Lona widget around a good drag and drop Javascript library

@maratori, @korantu, @laundmo, @themutt

**Comment by grigolet** [comment]
I really like the idea of all these events added to lona! Regarding drag and drop I don't have any direct experience with drag and drop libraries but looking around I found [interactjs](https://interactjs.io/) that has also for resizing events and multi touch gestures

**Reply by fscherf** [reply]
Interactjs looks very promising. Maybe the solution ist to just start ``lona-interactjs`` to add drag and drop etc

**Reply by grigolet** [reply]
I agree with you that given the high flexibility of Lona it could be easier to integrate interactsjs or any equivalent frontend library it as a separate package

**Comment by laundmo** [comment]
Keyboard events would be splendid, though i worry most usecases would have too many issues with latency.

**Reply by fscherf** [reply]
By keyboard events i meant something like shortcuts on an element like ``STRG + a`` or ``ARROW_UP`` to focus another element. Nothing for typing or fast movement really, but for short operations.

**Comment by fscherf** [comment]
FOCUS and BLUR are implemented now, available in the master branch, and will come with Lona 1.9.

@mwx23
